OverridenCore/CRM/Member/Form/Task/Email.php

```diff
@@ -0,0 +1,124 @@
+ +--------------------------------------------------------------------+
+ | This file is a part of CiviCRM.                                    |
+ |                                                                    |
+ | CiviCRM is free software; you can copy, modify, and distribute it  |
+ | under the terms of the GNU Affero General Public License           |
+ | Version 3, 19 November 2007 and the CiviCRM Licensing Exception.   |
+ |                                                                    |
+ | CiviCRM is distributed in the hope that it will be useful, but     |
+ | WITHOUT ANY WARRANTY; without even the implied warranty of         |
+ | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               |
+ | See the GNU Affero General Public License for more details.        |
+ |                                                                    |
+ | You should have received a copy of the GNU Affero General Public   |
+ | License and the CiviCRM Licensing Exception along                  |
+ | with this program; if not, contact CiviCRM LLC                     |
+ | at info[AT]civicrm[DOT]org. If you have questions about the        |
+ | GNU Affero General Public License or the licensing of CiviCRM,     |
+ | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC (c) 2004-2017
+ * $Id: Email.php 45499 2013-02-08 12:31:05Z kurund $
+ *
+ */
+
+/**
+ * This class provides the functionality to email a group of
+ * contacts.
+ */
+class CRM_Member_Form_Task_Email extends CRM_Member_Form_Task {
+
+  /**
+   * Are we operating in "single mode", i.e. sending email to one
+   * specific contact?
+   *
+   * @var boolean
+   */
+  public $_single = FALSE;
+
+  /**
+   * Are we operating in "single mode", i.e. sending email to one
+   * specific contact?
+   *
+   * @var boolean
+   */
+  public $_noEmails = FALSE;
+
+  /**
+   * All the existing templates in the system.
+   *
+   * @var array
+   */
+  public $_templates = NULL;
+
+  /**
+   * Build all the data structures needed to build the form.
+   *
+   * @return void
+   */
+  public function preProcess() {
+    CRM_Contact_Form_Task_EmailCommon::preProcessFromAddress($this);
+    parent::preProcess();
+
+    // we have all the membership ids, so now we get the contact ids
+    parent::setContactIDs();
+
+    $this->assign('single', $this->_single);
+  }
+
+  /**
+   * Build the form object.
+   *
+   *
+   * @return void
+   */
+  public function buildQuickForm() {
+    //enable form element
+    $this->assign('emailTask', TRUE);
+
+    CRM_Contact_Form_Task_EmailCommon::buildQuickForm($this);
+  }
+
+  /**
+   * Process the form after the input has been submitted and validated.
+   *
+   * @return void
+   */
+  public function postProcess() {
+    //start inject code
+    $messageTemplateId = $this->getVar('_submitValues')['template'];
```

Comment:
Same here, I'd think we would rather find some other solution than overriding core files. Please also explain what you need to do here, to check from our side if we can come up with some way to do it, other than overriding.

Reply:
If we go to the form “find contribution/ membership/activity” then selected some entity and click on “Email - send now” button”. Then redirected to “send email form”.
We override core, because has some issue in this form:

1. For this form, message template uses token for dynamic information. We can’t create token because, for creating a token(tokens create in two hooks: “hook_civicrm_tokens” and “hook_civicrm_tokenValues”) we don't have any data like contribution/membership/activity id, but we have only contacts id. That's why this way doesn't work. Also, CiviCRM has default tokens with some contribution data(it generated in core), but we can’t use it in  hooks for new tokens.

2. Smarty is disabled for default. We can enable smarty for this templates in settings(only in file ”civicrm.settings.php”). But smarty in templates can only use token variables.(you can see it in documentation: https://docs.civicrm.org/user/en/latest/common-workflows/tokens-and-mail-merge/#smarty ). 

3. We can’t collect data and put it into template.
4. Of course we can move logic to “postProcces” hook instead of override files, but it will be send two emails:
 - One broken (without collected data, and viewable smarty syntax).
 - One ours email with all data.
 That's why we didn’t do it.

In forms we override postProcces method. There is checked if a user chooses our message templates. And then it is collected all data and send to email.

Reply:
Thanks for the explanation. If this is to be used only on the actions though, maybe we could implement the links hook instead to rewrite those actions' URL's and use our own class. I think you'd only need to change those class' names and add them as callbacks for the new URL's. Of course, you'd also need to implement the template for the form, but maybe you could just include the CiviCRM core template and be done with it.

Let me know what you think.

Reply:
Before we override 3 form:
1.“CRM_Activity_Form_Task_Email”
2.“CRM_Contribute_Form_Task_Email”
3.“CRM_Member_Form_Task_Email”
And now we try to do it without overriding.

Form “find membership/activity/contribution”. There we can put new item in “action select” and run some attached form. It works good.
But then we create new forms extends core forms : 
1.CRM_ManualDirectDebit_Form_Email_Activity extends CRM_Activity_Form_Task_Email
2.CRM_ManualDirectDebit_Form_Email_Contribution extends 
3.CRM_Contribute_Form_Task_Email
CRM_ManualDirectDebit_Form_Email_Membership extends CRM_Member_Form_Task_Email

From “CRM_ManualDirectDebit_Form_Email_Contribution” works good.
And forms “CRM_ManualDirectDebit_Form_Email_Membership/Activity” is failed because this forms uses additional form between and its crash our data.

Maybe point leaving  it as it was with overridings. Because I  guess if we will follow this way it may not bring expected results.

Reply:
I'm afraid I'd still insist we try and avoid overriding, even if we need to implement our own way to handle all messages, but I'd like to know what @guanhuan and @omarabuhussein have to say about the matter. What do you guys think?